Mbed Crypto Tests

itayzafrir opened the pull request:
### Description

Add mbed-crypto basic functionality tests to mbed-os/TESTS directory.
<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [x] Test update
    [ ] Breaking change

Commit:
Introduce mbed-crypto tests

Add tests setup code and random generator test.

Commit:
Add mbed-crypto asymmetric encrypt decrypt test

Commit:
Add mbed-crypto hash verify test

Commit:
Add mbed-crypto symmetric cipher encrypt decrypt test

Commit:
Add mbed-crypto asymmetric sign verify test

Commit:
Add mbed-crypto key derivation test

Patater commented on TESTS/mbed-crypto/sanity/main.cpp:
```diff
@@ -51,7 +82,8 @@ utest::v1::status_t test_setup(const size_t number_of_cases)
 Case cases[] = {
-    Case("mbed-crypto random", test_crypto_random, case_failure_handler)
+    Case("mbed-crypto random", test_crypto_random, case_failure_handler),
+    Case("mbed-crypto asymmetric encrypt/decrypt", test_crypto_asymmetric_encrypt_decrypt, case_failure_handler)
```
Does it break anything to always include the comma at the end of the line here?

If we can always include commas at the end, the diffs will be more clean (won't look like we are adding new tests that were already added in a previous commit).

itayzafrir replied:
I'm not sure I'll check if it breaks. I can also add a line break before the closing `}` of the cases array instead, then it will prevent the false appearance.

itayzafrir replied:
I'll fix this once I raise an actual PR to mbed-os. I'll either rebase this branch or create a completely new branch.

Patater commented on TESTS/mbed-crypto/sanity/main.cpp:
```diff
@@ -68,6 +68,25 @@ void test_crypto_asymmetric_encrypt_decrypt(void)
+{
+    psa_algorithm_t alg = PSA_ALG_SHA_256;
+    psa_hash_operation_t operation;
+    unsigned char hash[] = {
```
What is this a hash of? Include a comment in the code.

itayzafrir replied:
OK will do.

itayzafrir replied:
Done

Patater replied:
OK

Patater commented on TESTS/mbed-crypto/sanity/main.cpp:
```diff
@@ -87,6 +87,44 @@ void test_crypto_hash_verify(void)
+                                                     encrypted, sizeof(encrypted), &output_len));
+    TEST_ASSERT_EQUAL(PSA_SUCCESS, psa_cipher_finish(&operation, encrypted + output_len,
+                                                     sizeof(encrypted) - output_len, &output_len));
+    TEST_ASSERT_EQUAL(PSA_SUCCESS, psa_cipher_abort(&operation));
```
We should add another TEST_ASSERT_EQUAL_UINT8_ARRAY here, to check that the encrypted value matches what we expect the data to get encrypted to.

Otherwise, this test would pass if the buffer was never encrypted.

itayzafrir replied:
I know, but since the key is generated rather than imported I don't see how this can be done. The jira tasks states that the keys should be generated.

Patater replied:
We have a separate key generation test. Not all tests must use generated keys.

itayzafrir replied:
I'll change this to use a fixed imported key.

itayzafrir replied:
Changed to deterministic.

Patater replied:
OK

Patater commented on TESTS/mbed-crypto/sanity/main.cpp:
```diff
@@ -125,6 +125,34 @@ void test_crypto_symmetric_cipher_encrypt_decrypt(void)
+    psa_algorithm_t alg = PSA_ALG_RSA_PKCS1V15_CRYPT;
+    size_t key_bits = 512;
+    psa_key_policy_t policy;
+    unsigned char input[100];
```
Could we test with a fixed input so we can check the result of the signature matches what we expect? Note that this would only work for deterministic signature algorithms.

itayzafrir replied:
Same comment as before, the key is generated.

Patater replied:
The signature test doesn't need to use a generated key.

itayzafrir replied:
Ok I'll change this to use a fixed imported key then.

itayzafrir replied:
Changed to deterministic.

Commit:
Add inline documentation

Commit:
Fix test test_crypto_asymmetric_sign_verify

Add the test to the test cases array, this was missed previously.
Fix the buffer sizes in the test.

Commit:
Change symmetric cipher encrypt decrypt test to be deterministic

Commit:
Change symmetric cipher encrypt decrypt test to be deterministic

itayzafrir reviewed (commented):
@Patater I've updated the following tests to be deterministic:
* `test_crypto_symmetric_cipher_encrypt_decrypt` 
* `test_crypto_asymmetric_sign_verify`

The test `test_crypto_asymmetric_encrypt_decrypt` remains as is for now, i.e. uses a generated key and is non-deterministic. Please let me know if you want this one to be deterministic as well.

I've also addressed the rest of the comments in your review.

I haven't rebased the branch to keep the review process easier, once approved I'll either rebase and get rid of some of the commits or create an new branch altogether.

alekshex commented on TESTS/mbed-crypto/sanity/main.cpp:
```diff
@@ -0,0 +1,301 @@
+        0x28, 0x8d, 0x76, 0xc0, 0xa7, 0x09, 0x50, 0x3f,
+        0x87, 0x96, 0x1e, 0x96, 0x05, 0xcb, 0xb9, 0x6d
+    };
+    unsigned char encrypted[sizeof(input)], decrypted[sizeof(input)], iv[sizeof(input)];
```
@itayzafrir  is decrypted buffer size sizeof(input) or can be size of output, which is bigger?

itayzafrir replied:
The input buffer is 16 bytes (one block as this is AES), and since we're using `PSA_ALG_CBC_NO_PADDING` they should all be the same size.

mohammad1603 commented on TESTS/mbed-crypto/sanity/main.cpp:
```diff
@@ -0,0 +1,301 @@
+        0x28, 0x8d, 0x76, 0xc0, 0xa7, 0x09, 0x50, 0x3f,
+        0x87, 0x96, 0x1e, 0x96, 0x05, 0xcb, 0xb9, 0x6d
+    };
+    unsigned char encrypted[sizeof(input)], decrypted[sizeof(input)], iv[sizeof(input)];
```
In case the `input` buffer will be changed in the future to larger one,  this will cause to invalid `IV` size.
I think it should be hard-coded to 16.

itayzafrir replied:
Done!

Commit:
Hard code IV size in test_crypto_symmetric_cipher_encrypt_decrypt

Patater commented on TESTS/mbed-crypto/sanity/main.cpp:
```diff
@@ -0,0 +1,301 @@
+    TEST_ASSERT_EQUAL(PSA_SUCCESS, psa_generate_key(slot, key_type, key_bits, NULL, 0));
+    TEST_ASSERT_EQUAL(PSA_SUCCESS, psa_get_key_information(slot, NULL, &got_bits));
+    TEST_ASSERT_EQUAL(key_bits, got_bits);
+    TEST_ASSERT_EQUAL(PSA_SUCCESS, psa_asymmetric_encrypt(slot, alg, input, sizeof(input), NULL, 0,
```
I think we've talked about this before, but "encrypt and then decrypt" is a bad example, as nobody will do that in practice. Encrypt should be an example. Decrypt should be a separate example.

This applies to both asymmetric and symmetric operations.

itayzafrir replied:
I agree for an example, but this is a test. Our test suite also has tests which encrypt and decrypt:

https://github.com/ARMmbed/mbedtls-psa/blob/c11be0943e0700526df95071208de0fd2d063baf/tests/suites/test_suite_psa_crypto.function#L2485

https://github.com/ARMmbed/mbedtls-psa/blob/c11be0943e0700526df95071208de0fd2d063baf/tests/suites/test_suite_psa_crypto.function#L244

Patater replied:
OK, fine to make our tests better later.

Patater commented on TESTS/mbed-crypto/sanity/main.cpp:
```diff
@@ -0,0 +1,301 @@
+    psa_key_type_t key_type = PSA_KEY_TYPE_RSA_KEYPAIR;
+    psa_algorithm_t alg = PSA_ALG_RSA_PKCS1V15_SIGN_RAW;
+    psa_key_policy_t policy;
+    unsigned char key[] = {
```
Consider making this, and other large constant arrays, `static const`.

itayzafrir replied:
What do you consider a **large** array?

Patater replied:
What do you consider to be excessive stack usage for an embedded target?

itayzafrir replied:
Changed all constant arrays to `static const`

Patater replied:
OK

Commit:
Change all constant arrays to static const

Patater reviewed (approved):
OK

itayzafrir commented:
New PR @ https://github.com/ARMmbed/mbed-os/pull/8754
